external/fv3fit/fv3fit/pytorch/cyclegan/generator.py

```diff
@@ -128,30 +129,36 @@ def up(in_channels: int, out_channels: int):
 
         min_filters = int(max_filters / 2 ** n_convolutions)
 
-        self._first_conv = nn.Sequential(
+        first_conv = nn.Sequential(
+            FoldTileDimension(nn.ReflectionPad2d(3)),
             convolution(
                 kernel_size=7,
                 in_channels=channels,
                 out_channels=min_filters,
-                padding="same",
+                padding=0,
             ),
+            FoldTileDimension(nn.InstanceNorm2d(min_filters)),
             relu_activation()(),
         )
 
-        self._encoder_decoder = SymmetricEncoderDecoder(
+        encoder_decoder = SymmetricEncoderDecoder(
             down_factory=down,
             up_factory=up,
             bottom_factory=resnet,
             depth=n_convolutions,
             in_channels=min_filters,
         )
 
-        self._out_conv = convolution(
-            kernel_size=7,
-            in_channels=min_filters,
-            out_channels=channels,
-            padding="same",
+        out_conv = nn.Sequential(
+            FoldTileDimension(nn.ReflectionPad2d(3)),
```

Comment:
I'm not sure yet. One possibility is that this module becomes some kind of HaloUpdate module, and we still use normal convolution. Another is that we use graph convolution or message passing, and remove this module. I'll likely try a few things and see what works best.

I am a little surprised by the thought that this only needs 3 halo (reflection padded) points, when the receptive field is definitely more than 3 points. I guess it's that the input/output layers need to have padded input because the definition of their input is static (it's real data), whereas the latent variables can be defined in a way that zero-padding (which is what it does) works just fine. This might mean a halo update based approach isn't a good idea.